src/cpp/fastdds/subscriber/DataReaderImpl.cpp

```diff
@@ -218,12 +225,412 @@ DataReaderImpl::~DataReaderImpl()
     delete user_datareader_;
 }
 
+bool DataReaderImpl::can_be_deleted() const
+{
+    if (reader_ != nullptr)
+    {
+        std::lock_guard<RecursiveTimedMutex> lock(reader_->getMutex());
+        return !loan_manager_.has_outstanding_loans();
+    }
+
+    return true;
+}
+
 bool DataReaderImpl::wait_for_unread_message(
         const fastrtps::Duration_t& timeout)
 {
     return reader_ ? reader_->wait_for_unread_cache(timeout) : false;
 }
 
+ReturnCode_t DataReaderImpl::check_collection_preconditions_and_calc_max_samples(
+        LoanableCollection& data_values,
+        SampleInfoSeq& sample_infos,
+        int32_t& max_samples)
+{
+    // Properties should be the same on both collections
+    if ((data_values.has_ownership() != sample_infos.has_ownership()) ||
+            (data_values.maximum() != sample_infos.maximum()) ||
+            (data_values.length() != sample_infos.length()))
+    {
+        return ReturnCode_t::RETCODE_PRECONDITION_NOT_MET;
+    }
+
+    // Check if a loan is required
+    if (0 < data_values.maximum())
+    {
+        // Loan not required, input collections should not be already loaned
+        if (false == data_values.has_ownership())
+        {
+            return ReturnCode_t::RETCODE_PRECONDITION_NOT_MET;
+        }
+
+        uint32_t collection_input_max = data_values.maximum();
+        int32_t collection_max = (collection_input_max > std::numeric_limits<uint32_t>::max() / 2u) ?
+                std::numeric_limits<int32_t>::max() : static_cast<int32_t>(collection_input_max);
```

Comment:
Missing tests for the second clause

Reply:
I would say the missing test is for the first one (when the input collection has a maximum value above the maximum of a signed integer).

Reply:
Adding a test for this would mean using a collection with space for 2^31 items. It may use up 520 GB of memory.

Reply:
I have just thought of a way to check this without using a huge amount of memory. Will implement it after lunch.